README
search.cpan.org clone based on MetaCPAN API

This is an attempt to recreate the look-and-feel of http://search.cpan.org/ using the
MetaCPAN API as a back-end. https://github.com/CPAN-API/cpan-api/

For objectives, plans, and a description of what the site, and the code behind it should look like see
http://perlmaven.com/search-cpan-org

For a live version of this application visit http://sco.perlmaven.com/


Development environment:
------------------------

* Install the prerequisites via Makefile.PL
* Run bin/fetch.pl to fetch some meta-meta-data from MetaCPAN::API
* Run plackup -r
* Borwse to http://localhost:5000/



Changes

0.01
    Start working on the application


t/lib/Test.pm
package t::lib::Test;
use strict;
use warnings;

use Exporter qw(import);
use Test::Builder;
use HTML::Tidy;

our @EXPORT = qw(contains html_check html_tidy);

sub html_tidy {
	my $tidy = HTML::Tidy->new;

	# HTML 4: <script src="/jquery.js" type="text/javascript"></script>
	# HTML 5: <script src="/jquery.js"></script>
	$tidy->ignore( text => qr{<script> inserting "type" attribute} );

	# HTML 4: <link rel="stylesheet" href="/style.css" type="text/css" />
	# HTML 5: <link rel="stylesheet" href="/style.css" />
	$tidy->ignore( text => qr{<link> inserting "type" attribute} );

	# HTML 4.01    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
	# HTML 5       <meta charset="utf-8" />
	$tidy->ignore( text => qr{<meta> proprietary attribute "charset"} );
	$tidy->ignore( text => qr{<meta> lacks "content" attribute} );

	# AFAIK HTML 5 does not support the "summary" attribute
	$tidy->ignore( text => qr{<table> lacks "summary" attribute} );

	# We should probably replace & in gravatar URLS by &amp; instead of hiding the warning:
	#$tidy->ignore( text => qr{unescaped & or unknown entity "&d"} );

	#$tidy->ignore( text => qr{inserting} ); #: <script> inserting "type" attribute} );

	return $tidy;
}



sub contains {
	my ( $str, $expected, $name ) = @_;
	$name //= '';

	my $T = Test::Builder->new;
	my $res = $T->ok(0 < index( $str, $expected ), $name);
	if ( not $res ) {
		$T->diag($str);
		$T->diag("\nDoes not contain:\n\n");
		$T->diag("'$expected'");
	}
	return $res;
}

sub html_check {
	my ( $html, $name ) = @_;
	$name //= 'html_check';

	my $T = Test::Builder->new;
	my @rows = split /\r?\n/, $html;
	my @fails;
	foreach my $i ( 0 .. @rows - 1 ) {

		#if ( $rows[$i] =~ /class=(?!")/ ) {
		if ( $rows[$i] =~ m{<\w+(\s+\w+="[^"]*")*\s+\w+=\w} ) {
			push @fails, "row $i   $rows[$i]";
		}
	}
	$T->ok( @fails == 0, $name );
	foreach my $f (@fails) {
		$T->diag($f);
	}
}

1;



Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

my %conf = (
	NAME         => 'MetaCPAN::SCO',
	AUTHOR       => 'Gabor Szabo <szabgab@cpan.org>',
	VERSION_FROM => 'lib/MetaCPAN/SCO.pm',
	PREREQ_PM    => {
		'Cwd'                    => '0',
		'Data::Dumper'           => '0',
		'File::Basename'         => '0',
		'HTTP::Tiny'             => '0',
		'Plack::Request'         => '0',
		'Template'               => '0',
		'Template::Plugin::Page' => '0',
		'LWP::Simple'            => '0',
		'JSON'                   => '0',
		'Path::Tiny'             => '0',
		'Pod::Simple::HTML'      => '0',
		'POSIX'                  => '0',
	},
);

if ( eval { ExtUtils::MakeMaker->VERSION(6.3002) } ) {
	$conf{LICENSE} = 'perl';
}

if ( eval { ExtUtils::MakeMaker->VERSION(6.46) } ) {
	$conf{META_MERGE} = {
		'meta-spec' => { version => 2 },
		resources   => {
			repository => {
				type    => 'git',
				url     => 'http://github.com/szabgab/MetaCPAN-SCO.git',
				web     => 'http://github.com/szabgab/MetaCPAN-SCO',
				license => 'http://dev.perl.org/licenses/',
			},
			bugtracker => {
				web => 'http://github.com/szabgab/MetaCPAN-SCO/issues',
			},
			homepage => 'http://perlmaven.com/',
		},
	};
}

my %configure_requires = ( 'ExtUtils::MakeMaker' => '6.64', );
my %build_requires     = ();
my %test_requires      = (
	'Test::More'            => '1.00',
	'Test::WWW::Mechanize'  => '0',
	'Test::Code::TidyAll'   => '0',
	'Perl::Tidy'            => '0',
	'Perl::Critic'          => '0',
	'Test::Perl::Critic'    => '0',
	'Plack::Test'           => '0',
	'HTTP::Request::Common' => '0',
	'Test::HTML::Tidy'      => '0',
	'HTML::Tidy'            => '0',

);

###   merging data "standard code"
if ( eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
	$conf{CONFIGURE_REQUIRES} = \%configure_requires;
}
else {
	%{ $conf{PREREQ_PM} } = ( %{ $conf{PREREQ_PM} }, %configure_requires );
}

if ( eval { ExtUtils::MakeMaker->VERSION(6.5503) } ) {
	$conf{BUILD_REQUIRES} = \%build_requires;
}
else {
	%{ $conf{PREREQ_PM} } = ( %{ $conf{PREREQ_PM} }, %build_requires );
}
if ( eval { ExtUtils::MakeMaker->VERSION(6.64) } ) {
	$conf{TEST_REQUIRES} = \%test_requires;
}
else {
	%{ $conf{PREREQ_PM} } = ( %{ $conf{PREREQ_PM} }, %test_requires );
}

WriteMakefile(%conf);



lib/MetaCPAN/SCO/Fetch.pm
package MetaCPAN::SCO::Fetch;
use strict;
use warnings;

use LWP::Simple qw(get);
use JSON qw(from_json to_json);
use Path::Tiny qw(path);

=pod

TODO: Once we are sure these are the correct queries, add them to https://github.com/CPAN-API/cpan-api/wiki/API-docs

curl -XPOST api.metacpan.org/v0/author/_search -d '{
  "query": { "match_all": {} },
  "size":0
}'
curl http://api.metacpan.org/v0/author/_search?size=0


curl -XPOST api.metacpan.org/v0/distribution/_search -d '{
  "query": { "match_all": {} },
  "size":0
}'
curl http://api.metacpan.org/v0/distribution/_search?size=0


curl -XPOST api.metacpan.org/v0/module/_search -d '{
  "query": { "match_all": {} },
  "size":0
}'
curl http://api.metacpan.org/v0/module/_search?size=0

=cut

sub run {
	my ( $self, $root ) = @_;

	my %totals;
	foreach my $name (qw(author distribution module)) {
		my $json = get "http://api.metacpan.org/v0/$name/_search?size=0";
		my $data = from_json $json;
		$totals{$name} = $data->{hits}{total};
	}
	path("$root/totals.json")->spew_utf8( to_json \%totals );
	return;
}

1;



lib/MetaCPAN/SCO.pm
package MetaCPAN::SCO;
use strict;
use warnings;

use Carp ();
use Cwd qw(abs_path);
use Data::Dumper qw(Dumper);
use File::Basename qw(dirname);
use HTTP::Tiny;
use JSON qw(from_json to_json);
use LWP::Simple qw(get);
use Path::Tiny qw(path);
use Plack::Builder;
use Plack::Response;
use Plack::Request;
use Pod::Simple::HTML;
use POSIX qw(strftime);
use Template;
use Time::Local qw(timegm);

our $VERSION = '0.01';

=head1 NAME

SCO - search.cpan.org clone

=cut

my $env;

sub run {
	my $root = root();

	my $app = sub {
		$env = shift;

		my $request   = Plack::Request->new($env);
		my $path_info = $request->path_info;
		if ( $path_info eq '/' ) {
			return template(
				'index',
				{
					front => 1,
					title => 'The CPAN Search Site - search.cpan.org',
				}
			);
		}
		if ( $path_info eq '/feedback' ) {
			return template( 'feedback',
				{ title => 'Site Feedback - search.cpan.org', } );
		}
		if ( $path_info eq '/faq.html' ) {
			return template( 'faq', { title => 'FAQ - search.cpan.org', } );
		}
		if ( $path_info eq '/recent' ) {
			my $recent = recent( $request->param('d') );
			return template( 'recent', { recent => $recent } );
		}
		if ( $path_info =~ m{^/author/?$} ) {
			my $query_string = $request->query_string;
			return template(
				'authors',
				{
					letters => [ 'A' .. 'Z' ],
					authors => [],
					title   => 'The CPAN Search Site - search.cpan.org',
				}
			) if not $query_string;
			my $lead = substr $query_string, 0, 1;
			my $authors = authors_starting_by( uc $lead );
			if (@$authors) {
				return template(
					'authors',
					{
						letters         => [ 'A' .. 'Z' ],
						authors         => $authors,
						selected_letter => uc($lead),
						title => 'The CPAN Search Site - search.cpan.org',
					}
				);
			}
		}

		if ( $path_info =~ m{^/~([a-z]+)$} ) {
			my $res = Plack::Response->new();
			$res->redirect( "$path_info/", 301 );
			return $res->finalize;
		}
		if ( $path_info =~ m{^/~([a-z]+)/$} ) {
			my $pauseid = uc $1;
			my $author  = get_author_info($pauseid);
			$author->{cpantester} = substr( $pauseid, 0, 1 ) . '/' . $pauseid;
			my $distros = get_distros_by_pauseid($pauseid);
			return template(
				'author',
				{
					author        => $author,
					distributions => $distros,
					title         => "$author->{name}  - search.cpan.org",
				}
			);
		}

		# ~ealleniii/Config-Options-0.08/
		if ( $path_info =~ m{^/~([a-z]+)/([^/]+)/(.*)?$} ) {
			my ( $pauseid, $dist_name, $file ) = ( uc($1), $2, $3 );
			if ( not $file ) {
				my $data = get_dist_data( $pauseid, $dist_name );

				return template( 'dist', $data );
			}

			if ( $file eq 'MANIFEST' ) {
				my $manifest = get
					"http://api.metacpan.org/source/$pauseid/$dist_name/$file";
				my @rows = split /\r?\n/, $manifest;
				my @entries;
				foreach my $row (@rows) {
					next if $row =~ /^\s*$/;
					$row =~ s/^\s+|\s+$//g;
					my ( $file, $text ) = split /\s+/, $row, 2;
					my %e = (
						file => $file,
						text => $text,
					);
					if ( $file =~ /\.(pod|pm)$/ ) {
						$e{pod} = $file;
					}
					push @entries, \%e;
				}
				my $author = get_author_info($pauseid);
				return template(
					'manifest',
					{
						manifest  => \@entries,
						pauseid   => $pauseid,
						dist_name => $dist_name,
						author    => $author,
						username  => lc($pauseid)
					}
				);
			}

			if ( $file =~ /\.(pod|pm)$/ ) {
				my $source = get
					"http://api.metacpan.org/source/$pauseid/$dist_name/$file";

				my $p = Pod::Simple::HTML->new;
				$p->output_string( \my $pod );
				$p->index(1);
				$p->html_header_before_title('');
				$p->html_header_after_title('');
				$p->html_footer('');
				$p->parse_string_document($source);

		   # fetch info about the distribution to show on the right-hand side.

				return template( 'pod', { pod => $pod } );
			}

		}

		if ( $path_info =~ m{^/dist/([^/]+)$} ) {
			my $res = Plack::Response->new();
			$res->redirect( "$path_info/", 301 );
			return $res->finalize;
		}

		if ( $path_info =~ m{^/dist/([^/]+)/$} ) {
			my $dist_name      = $1;
			my $latest_release = get_latest_release($dist_name);
			my $data           = get_dist_data( $latest_release->{author},
				$latest_release->{name} );
			return template( 'dist', $data );
		}

		#if ($path_info =~ m{^/src/([^/]+)/([^/]+)/(.*)}) {
		if ( $path_info =~ m{^/src/(.*)} ) {
			my $res = Plack::Response->new();

	#http://api.metacpan.org/source/DDUMONT/Config-Model-Itself-1.241/Build.PL

# meta information about a file:
# http://api.metacpan.org/v0/file/_search?q=path:Build.PL%20AND%20release:Config-Model-Itself-1.241&size=1
			$res->redirect( "http://api.metacpan.org/source/$1", 301 );
			return $res->finalize;
		}

		if ( $path_info eq '/search' ) {

# I found out that the pager on sco uses the paramters q,m,s,n
# and there is actually a strange issue. The pager links on the search result page show numbers 1,2,3,4 (the page numbers)
# but the links underneath use s=<the number of the first hit on that page>
# the user can supply other values to s=, not only the onese that are the starting points of "real" pages. for example s=13
# In that case the new page will show the correct number of results starting from result #13 but the links will stull say
# pages 1,2,3 ...
# It might be better to just use p= page number to send to the server
			my $query = $request->param('query') || $request->param('q');
			my $mode  = $request->param('mode')  || $request->param('m');

			#my $start = $request->param('s') || 1;
			my $page = $request->param('p') || 1;
			my $size = $request->param('n') || 10;
			return search( $query, $mode, $page, $size );
		}

		my $reply = template('404');
		return [ '404', [ 'Content-Type' => 'text/html' ], $reply->[2], ];
	};

	builder {
		enable 'Plack::Middleware::Static',
			path => qr{^/(favicon.ico|robots.txt)},
			root => "$root/static/";
		$app;
	};
}

sub get_latest_release {
	my ($dist_name) = @_;

	my $json = get "http://api.metacpan.org/v0/release/$dist_name";
	my $data = from_json $json;
	return $data;
}

sub get_releases {
	my ($dist_name) = @_;

	my $json = get
		"http://api.metacpan.org/v0/release/_search?q=distribution:$dist_name&size=30&fields=author,name,date,status,abstract";
	my $data = from_json $json;
	my @releases = reverse sort { $a->{date} cmp $b->{date} }
		grep { $_->{status} eq 'cpan' }
		map  { $_->{fields} } @{ $data->{hits}{hits} };
	return @releases;
}

sub get_dist_data {
	my ( $pauseid, $dist_name_ver ) = @_;

	# curl 'http://api.metacpan.org/v0/release/AADLER/Games-LogicPuzzle-0.20'
	# curl 'http://api.metacpan.org/v0/release/Games-LogicPuzzle'
	# from https://github.com/CPAN-API/cpan-api/wiki/API-docs
	my $dist;
	my $release;
	my @files;

	eval {
		my $json1
			= get 'http://api.metacpan.org/v0/release/'
			. $pauseid . '/'
			. $dist_name_ver;
		$dist = from_json $json1;
		my $json2 = get
			"http://api.metacpan.org/v0/file/_search?q=release:$dist_name_ver&size=1000&fields=release,path,module.name,abstract,module.version,documentation,directory";
		my $data2 = from_json $json2;
		@files = map { $_->{fields} } @{ $data2->{hits}{hits} };

		1;
	} or do {
		my $err = $@ // 'Unknown error';
		warn $err if $err;
	};
	my @releases = get_releases( $dist->{metadata}{name} );

	my %SPECIAL = map { $_ => 1 } qw(
		Changes CHANGES Changelog
		LICENSE MANIFEST README
		INSTALL
		Makefile.PL Build.PL META.yml META.json
		ARTISTIC SIGNATURE
	);

	for my $f (@files) {
		$f->{name}    = delete $f->{'module.name'};
		$f->{version} = delete $f->{'module.version'};

# If a file has several packages in it, then the 'name' field will be an ARRAY
# http://search.cpan.org/~perlancar/Locale-Tie-0.03/
		if ( ref $f->{name} ) {
			$f->{name} = $f->{name}[0];
		}
	}

	my @modules
		= sort { $a->{name} cmp $b->{name} } grep { $_->{name} } @files;
	my @documentation = sort { $a->{documentation} cmp $b->{documentation} }
		grep {
		        $_->{documentation}
			and not $_->{name}
			and $_->{path} !~ m{^t/}
		} @files;

# It seem sco shows META.json if it is available or META.yml if that is available, but not both
# and prefers to show META.json
# http://search.cpan.org/~jdb/PPM-Repositories-0.20/
# http://search.cpan.org/~ironcamel/Business-BalancedPayments-1.0401/
# I wonder if showing both (when available) can be considered a slight improvement or if we should
# hide META.yml if there is a META.json already?

	my %special
		= map { $_->{path} => $_ } grep { $SPECIAL{ $_->{path} } } @files;
	if ( $special{'META.json'} ) {
		delete $special{'META.yml'};
	}

	my @other_files = sort { lc $a->{path} cmp lc $b->{path} }
		grep {
		        not $SPECIAL{ $_->{path} }
			and not $_->{documentation}
			and not $_->{name}
			and not $_->{directory} eq 'true'

# TODO: unclear why to filter these but they were not shown on http://search.cpan.org/~tlinden/apid-0.04/
# README.md, sample/index.html sample/README
#and not( $_->{path} =~ m{(\.map|\.conf|\.ini|cpanfile)$} )
#and not( $_->{path} =~ m{t/} )
# TODO: http://search.cpan.org/dist/HTML-Template/  has all kinds of other files to filter, so there probably need to be a white-list
			and ( $_->{path} =~ /README/ or $_->{path} =~ m{\.html$} )
		} @files;

# TODO: the MANIFEST file gets special treatment here and instead of linking to src/ it is linked without
# anything and then it is shown with links to the actual files.
	my @special_files
		= sort { lc $a->{path} cmp lc $b->{path} } values %special;
	$dist->{this_name} = $dist->{name};
	my $author = get_author_info($pauseid);
	return {
		dist          => $dist,
		author        => $author,
		special_files => \@special_files,
		modules       => \@modules,
		documentation => \@documentation,
		releases      => \@releases,
		other_files   => \@other_files,
		title         => "$author->{name} / $dist->{name} - search.cpan.org",
	};
}

sub recent {
	my ($end_ymd) = @_;

# http://api.metacpan.org/v0/release/_search?q=status:latest&fields=name,status,date&sort=date:desc&size=100

# 10 most recent releases by OALDERS
# curl 'http://api.metacpan.org/v0/release/_search?q=status:latest%20AND%20author:OALDERS&fields=name,author,status,date,abstract&sort=date:desc&size=10'

	$end_ymd //= strftime( '%Y%m%d', gmtime );
	my @ymd = unpack( 'A4 A2 A2', $end_ymd );
	my $end_y_m_d = join '-', @ymd;
	my $end_time    = timegm( 0, 0, 0, $ymd[2], $ymd[1] - 1, $ymd[0] );
	my $start_time  = $end_time - 7 * 24 * 60 * 60;
	my $start_y_m_d = strftime( '%Y-%m-%d', gmtime($start_time) );
	my $start_ymd   = strftime( '%Y%m%d', gmtime($start_time) );
	my $next_ymd;

	if ( $end_time < time - 24 * 60 * 60 ) {
		my $next_time = $end_time + 7 * 24 * 60 * 60;
		$next_ymd = strftime( '%Y%m%d', gmtime($next_time) );
	}

	my $ua         = HTTP::Tiny->new();
	my $query_json = to_json {
		query => {
			match_all => {},
		},
		filter => {
			and => [
				{ term => { status => 'latest', } },
				{
					range => {
						date => {
							from => "${start_y_m_d}T23:59:59",
							to   => "${end_y_m_d}T23:59:59",
						},
					},
				},
			]
		},
		fields => [qw(name author status date abstract)],
		sort   => { date => 'desc' },
		size   => 2000,
	};

	my @days;
	eval {
		my $resp = $ua->request(
			'POST',
			'http://api.metacpan.org/v0/release/_search',
			{
				headers => { 'Content-Type' => 'application/json' },
				content => $query_json,
			}
		);
		die if not $resp->{success};
		my $json    = $resp->{content};
		my $data    = from_json $json;
		my @distros = map { $_->{fields} } @{ $data->{hits}{hits} };
		my %days;
		foreach my $d (@distros) {
			push @{ $days{ substr( $d->{date}, 0, 10 ) } }, $d;
		}
		@days = map { { date => "${_}T12:00:00", dists => $days{$_} } }
			reverse sort keys %days;
	} or do {
		my $err = $@ // 'Unknown error';
		warn $err if $err;
	};
	my %resp = ( days => \@days, prev => $start_ymd );
	$resp{next} = $next_ymd;
	return \%resp;
}

sub search {
	my ( $query, $mode, $page, $page_size ) = @_;

	if ( $mode eq 'author' ) {
		my @authors = [];
		eval {
			my $json
				= get
				"http://api.metacpan.org/v0/author/_search?q=author.name:*$query*&size=5000&fields=name,asciiname,pauseid";
			my $data = from_json $json;
			@authors = sort { $a->{pauseid} cmp $b->{pauseid} }
				map { $_->{fields} } @{ $data->{hits}{hits} };
			1;
		} or do {
			my $err = $@ // 'Unknown error';
			warn $err if $err;
		};
		return template('no_matches') if not @authors;

		return template(
			'search_author',
			{
				authors      => \@authors,
				page_size    => $page_size,
				current_page => $page,
				mode         => $mode,
				query        => $query,
			}
		);
	}

	if ( $mode eq 'dist' ) {
		my @releases = [];
		eval {
			my $json
				= get
				"http://api.metacpan.org/v0/release/_search?q=name:*$query*&size=500&fields=date,name,author,abstract,distribution";
			my $data = from_json $json;
			@releases = sort { $a->{name} cmp $b->{name} }
				map { $_->{fields} } @{ $data->{hits}{hits} };

			1;
		} or do {
			my $err = $@ // 'Unknown error';
			die $err if $err;
		};

		return template('no_matches') if not @releases;
		return template(
			'search_dist',
			{
				dists        => \@releases,
				page_size    => $page_size,
				current_page => $page,
				mode         => $mode,
				query        => $query,
			}
		);
	}

	if ( $mode eq 'module' ) {
	}

	# 'all' is the default behaviour:

	return template('no_matches');
}

sub get_distros_by_pauseid {
	my ($pause_id) = @_;

 # curl 'http://api.metacpan.org/v0/release/_search?q=author:SZABGAB&size=500'
 # TODO the status:latest filter should be on the query not in the grep

	my @data;
	eval {
		my $json
			= get 'http://api.metacpan.org/v0/release/_search?q=author:'
			. $pause_id
			. '&size=500';
		my $raw = from_json $json;
		@data = sort { $a->{name} cmp $b->{name} }
			map {
			{
				name         => $_->{_source}{name},
				abstract     => $_->{_source}{abstract},
				date         => $_->{_source}{date},
				download_url => $_->{_source}{download_url},
			}
			}
			grep { $_->{_source}{status} eq 'latest' }
			@{ $raw->{hits}{hits} };
		1;
	} or do {
		my $err = $@ // 'Unknown error';
		warn $err if $err;
	};
	return \@data;
}

sub get_author_info {
	my ($pause_id) = @_;

 # The data is received from MetaCPAN it is not what the authors set on PAUSE.
 # See https://github.com/CPAN-API/cpan-api/issues/351
 # The source of the data on SCO is this xml file:
 # http://www.cpan.org/authors/00whois.xml

	my $data;
	eval {
		my $json
			= get 'http://api.metacpan.org/v0/author/_search?q=author._id:'
			. $pause_id
			. '&size=1';
		my $raw = from_json $json;
		$data = $raw->{hits}{hits}[0]{_source};
		1;
	} or do {
		my $err = $@ // 'Unknown error';
		warn $err if $err;
	};
	$data->{gravatar_url} =~ s{&}{&amp;}g;
	return $data;
}

sub authors_starting_by {
	my ($char) = @_;

# curl http://api.metacpan.org/v0/author/_search?size=10
# curl 'http://api.metacpan.org/v0/author/_search?q=author._id:S*&size=10'
# curl 'http://api.metacpan.org/v0/author/_search?size=10&fields=name'
# curl 'http://api.metacpan.org/v0/author/_search?q=author._id:S*&size=10&fields=name'
# or maybe use fetch to download and keep the full list locally:
# http://api.metacpan.org/v0/author/_search?pretty=true&q=*&size=100000 (from https://github.com/CPAN-API/cpan-api/wiki/API-docs )

	my @authors = [];
	if ( $char =~ /[A-Z]/ ) {
		eval {
			my $json
				= get
				"http://api.metacpan.org/v0/author/_search?q=author._id:$char*&size=5000&fields=name";
			my $data = from_json $json;
			@authors = sort { $a->{id} cmp $b->{id} }
				map { { id => $_->{_id}, name => $_->{fields}{name} } }
				@{ $data->{hits}{hits} };
			1;
		} or do {
			my $err = $@ // 'Unknown error';
			warn $err if $err;
		};
	}
	return \@authors;
}

sub template {
	my ( $file, $vars ) = @_;
	$vars //= {};
	Carp::confess 'Need to pass HASH-ref to template()'
		if ref $vars ne 'HASH';

	my $root = root();

	my $ga_file = "$root/config/google_analytics.txt";
	if ( -e $ga_file ) {
		$vars->{google_analytics} = path($ga_file)->slurp_utf8 // '';
	}

	eval {
		$vars->{totals} = from_json path("$root/totals.json")->slurp_utf8;
	};

	my $request = Plack::Request->new($env);
	$vars->{query} //= $request->param('query');
	$vars->{mode}  //= $request->param('mode');

	my $tt = Template->new(
		INCLUDE_PATH => "$root/tt",
		INTERPOLATE  => 0,
		POST_CHOMP   => 1,
		EVAL_PERL    => 1,
		START_TAG    => '<%',
		END_TAG      => '%>',
		PRE_PROCESS  => 'incl/header.tt',
		POST_PROCESS => 'incl/footer.tt',
	);
	my $out;
	$tt->process( "$file.tt", $vars, \$out )
		|| Carp::confess $tt->error();
	return [ '200', [ 'Content-Type' => 'text/html' ], [$out], ];
}

sub root {
	my $dir = dirname( dirname( dirname( abs_path(__FILE__) ) ) );
	$dir =~ s{blib/?$}{};
	return $dir;
}

1;

